`QuakeBridgeVisitor` to handle `cphase`

1tnguyen opened the pull request:
<!--
Thanks for helping us improve CUDA Quantum!

⚠️ The pull request title should be concise and understandable for all.
⚠️ If your pull request fixes an open issue, please link to the issue.

Checklist:
- [ ] I have added tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Description
`cphase` is a native op (declared in `qubit_qis.h`) yet not being handled by `QuakeBridgeVisitor`.

Hence, this PR adds "`cphase`" to the code path handling "`r1`" and adds a unit test to verify the AST.

I should also note that "`p`" is not a valid intrinsic gate name in `qubit_qis.h` but is handled here by `QuakeBridgeVisitor`.

schweitzpgi commented on test/AST-Quake/cphase.cpp:
```diff
@@ -0,0 +1,46 @@
+struct CtrlR1 {
+  void operator()() __qpu__ {
+    cudaq::qubit q1, q2;
+    r1<cudaq::ctrl>(1.234, q1, q2);
```
Might as well add a `p` test here, eh?

1tnguyen replied:
I wanted to as well, but `p` is not 'declared' in `qubit_qis.h`, hence will trigger a compile error.
If you want, I can add `p` to `qubit_qis.h`. What do you think?

amccaskey replied:
There is an open question we've had here as to whether we want `p` or `cphase`. Early on we decided against `p` in favor of `r1`, and since we generate controlled operations via the single target operations, we might want to avoid `cphase` in favor of always using `r1<cudaq::ctrl>`. I want @boschmitt to chime in here. 

So it might be that instead of adding `cphase` to the bridge, we should have removed it from `qubit_qis.h`.

amccaskey replied:
This is how I was thinking too. We should remove the cphase function from qubit_qis.h. @1tnguyen what do you think? Would you want to back out those bridge changes and instead remove cphase from qubit_qis.h? How does this affect your work?

schweitzpgi replied:
+1

1tnguyen replied:
Thanks for the explanation @boschmitt! 


> what do you think? How does this affect your work?

@amccaskey Totally agreed with having a single `r1` op to make API cleaner/less overloaded.
Actually, this doesn't affect my work at all. I was writing some CUDAQ kernels and looking for this gate. A quick search indicated that it was valid (used in some other places/tests in the codebase), but not handled in the bridge. I've since changed my own code to use `r1<ctrl>` but figured that it might be worth submitting a PR :-)

> Would you want to back out those bridge changes and instead remove `cphase` from `qubit_qis.h`?

Definitely. I'll update this PR accordingly.

schweitzpgi commented:
Nice catch. This looks good.

1tnguyen commented:
@amccaskey I've updated the PR to remove `cphase` from `qubit_qis.h` as suggested. Also removed the corresponding entry in `QubitQIRExecutionManager.cpp` and updated a couple of unit tests (`cphase` -> `r1<ctrl>`)

Side note: I don't know whether we also want to get rid of the underlying `__quantum__qis__cphase` QIR function or not  (i.e., only use `__quantum__qis__r1__ctl`), hence didn't go any further.

amccaskey reviewed (approved):
Thanks @1tnguyen

amccaskey commented:
@1tnguyen as soon as #135 gets merged, you can rebase this one and pick up the spell-check fixes for `qubit_qis.h`, then we'll be able to merge.

1tnguyen commented:
> @1tnguyen as soon as #135 gets merged, you can rebase this one and pick up the spell-check fixes for `qubit_qis.h`, then we'll be able to merge.

Thanks @amccaskey . Will do.
